paimon-core/src/main/java/org/apache/paimon/deletionvectors/DeletionVectorIndexFileWriter.java

```diff
@@ -0,0 +1,132 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.deletionvectors;
+
+import org.apache.paimon.fs.FileIO;
+import org.apache.paimon.fs.Path;
+import org.apache.paimon.index.IndexFileMeta;
+import org.apache.paimon.options.MemorySize;
+import org.apache.paimon.table.BucketMode;
+import org.apache.paimon.utils.Pair;
+import org.apache.paimon.utils.PathFactory;
+import org.apache.paimon.utils.Preconditions;
+
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.paimon.deletionvectors.DeletionVectorsIndexFile.DELETION_VECTORS_INDEX;
+import static org.apache.paimon.deletionvectors.DeletionVectorsIndexFile.VERSION_ID_V1;
+import static org.apache.paimon.deletionvectors.DeletionVectorsIndexFile.calculateChecksum;
+
+/** Writer for deletion vector index file. */
+public class DeletionVectorIndexFileWriter {
+
+    private final PathFactory indexPathFactory;
+    private final FileIO fileIO;
+    private final boolean isWrittenToMulitFiles;
+    private final long targetSizeInBytes;
+
+    private boolean written = false;
+
+    private long writtenSizeInBytes = 0L;
+
+    public DeletionVectorIndexFileWriter(
+            FileIO fileIO,
+            PathFactory pathFactory,
+            BucketMode bucketMode,
+            MemorySize targetSizePerIndexFile) {
+        this.indexPathFactory = pathFactory;
+        this.fileIO = fileIO;
+        this.isWrittenToMulitFiles = BucketMode.BUCKET_UNAWARE == bucketMode;
+        this.targetSizeInBytes = targetSizePerIndexFile.getBytes();
+    }
+
+    public List<IndexFileMeta> write(Map<String, DeletionVector> input) throws IOException {
+        List<IndexFileMeta> result = new ArrayList<>();
+        SingleIndexFileWriter writer = createWriter();
+
+        for (Map.Entry<String, DeletionVector> entry : input.entrySet()) {
+            String dataFile = entry.getKey();
+            byte[] valueBytes = entry.getValue().serializeToBytes();
+            int currentSize = valueBytes.length;
+
+            if (isWrittenToMulitFiles
+                    && written
+                    && writtenSizeInBytes + currentSize > targetSizeInBytes) {
+                result.add(writer.closeWriter());
+                writer = createWriter();
+            }
+
+            writer.write(dataFile, valueBytes);
+            written = true;
+            writtenSizeInBytes += currentSize;
+        }
+        result.add(writer.closeWriter());
+        return result;
+    }
+
+    private SingleIndexFileWriter createWriter() throws IOException {
+        written = false;
+        writtenSizeInBytes = 0L;
+        return new SingleIndexFileWriter(fileIO, indexPathFactory.newPath());
+    }
+
+    static class SingleIndexFileWriter {
```

Comment:
private, and maybe can be an inner class instead of static.
In this way, you can inline `createWriter` to the `SingleIndexFileWriter` constructor.